Source/JavaScriptCore/dfg/DFGBackwardsPropagationPhase.cpp

```diff
@@ -402,15 +400,15 @@ class BackwardsPropagationPhase {
                 flags |= NodeBytecodeUsesAsNumber;
             if (!m_allowNestedOverflowingAdditions)
                 flags |= NodeBytecodeUsesAsNumber;
-            flags |= NodeBytecodeNeedsNaNOrInfinity;
+            flags |= NodeBytecodeNeedsNaNOrInfinity | NodeBytecodeObservesUnsigned;
             
             node->child1()->mergeFlags(flags);
             node->child2()->mergeFlags(flags);
             break;
         }
 
         case ArithClz32: {
-            flags &= ~(NodeBytecodeUsesAsNumber | NodeBytecodeNeedsNegZero | NodeBytecodeNeedsNaNOrInfinity | NodeBytecodeUsesAsOther | ~NodeBytecodePrefersArrayIndex);
+            flags &= ~(NodeBytecodeUsesAsNumber | NodeBytecodeNeedsNegZero | NodeBytecodeNeedsNaNOrInfinity | NodeBytecodeUsesAsOther | ~NodeBytecodePrefersArrayIndex | ~NodeBytecodeObservesUnsigned);
```

Comment:
I’m pretty sure this code is quite wrong, and already wrong before the patch. If you carefully evaluate what it does after the patch, it just sets flags to `0`! I think maybe we should be leaving both of the `~` out of the expression to do what we intend. But also we need the kind of testing that would show us this mistake.

Reply:
Whoa, great catch! I'll remove the bit I added in this patch, and then we'll have to investigate if the existing `~ NodeBytecodePrefersArrayIndex` is causing any issues.